anchor/message_sender/src/network.rs

```
@@ -0,0 +1,141 @@
+                            error!(?err, "Creating signed message failed!");
+                            return;
+                        }
+                    };
+                    sender.do_send(message);
+                },
+                SIGNER_NAME,
+            )
+            .unwrap_or_else(|e| warn!("Failed to send to processor: {}", e));
+    }
+
+    fn send(&self, message: SignedSSVMessage) {
+        let sender = self.clone();
+        self.processor
+            .urgent_consensus
+            .send_blocking(
+                move || {
+                    sender.do_send(message);
+                },
+                SENDER_NAME,
+            )
+            .unwrap_or_else(|e| warn!("Failed to send to processor: {}", e));
+    }
+}
+
+impl NetworkMessageSender {
+    pub fn new(
+        processor: processor::Senders,
+        network_tx: mpsc::Sender<(SubnetId, Vec<u8>)>,
+        private_key: Rsa<Private>,
+        database: watch::Receiver<NetworkState>,
+        operator_id: OperatorId,
+        subnet_count: usize,
+    ) -> Result<Arc<Self>, String> {
+        let private_key = PKey::from_rsa(private_key)
+            .map_err(|err| format!("Failed to create PKey from RSA: {err}"))?;
+        Ok(Arc::new(Self {
+            processor,
+            network_tx,
+            private_key,
+            database,
+            operator_id,
+            subnet_count,
+        }))
+    }
+
+    fn do_send(&self, message: SignedSSVMessage) {
+        let subnet = match self.determine_subnet(&message) {
+            Ok(subnet) => subnet,
+            Err(err) => {
+                error!(?err, "Unable to determine subnet for outgoing message");
+                return;
+            }
+        };
+        match self.network_tx.try_send((subnet, message.as_ssz_bytes())) {
+            Ok(_) => debug!(?subnet, "Successfully sent message to network"),
+            Err(TrySendError::Closed(_)) => warn!("Network queue closed (shutting down?)"),
```

This error seems more critical. Should we propagate it and shut down the system?

---

To propagate it, we would need a return queue to the sender, and the sender would need to wait for it, defeating the purpose of putting this on another thread in the first place.

---

I believe this is fine If we can guarantee that this error happens if and only if the system is already shutting down. Otherwise, the client can keep calling `sign_and_send` and the network never receives a msg as it fails silently, right?

---

We could also add this at the beginning of `sign_and_send`

```rust
 if self.network_tx.is_closed() {
            // return error
        }
```